modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4IncrementalRequestHandlingIT.java

```diff
@@ -0,0 +1,344 @@
+import org.elasticsearch.rest.RestChannel;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestHandler;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.RestResponse;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.test.ESIntegTestCase;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import static io.netty.handler.codec.http.HttpHeaderNames.CONTENT_LENGTH;
+import static io.netty.handler.codec.http.HttpHeaderNames.CONTENT_TYPE;
+import static io.netty.handler.codec.http.HttpHeaderValues.APPLICATION_JSON;
+import static io.netty.handler.codec.http.HttpMethod.POST;
+import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+
+@ESIntegTestCase.ClusterScope(numDataNodes = 1)
+public class Netty4IncrementalRequestHandlingIT extends ESNetty4IntegTestCase {
+
+    // ensure empty http content has single 0 size chunk
+    public void testEmptyContent() throws Exception {
+        try (var ctx = setupClientCtx()) {
+            ctx.clientChannel.writeAndFlush(fullRequest(Unpooled.EMPTY_BUFFER));
+            var handler = ctx.awaitRestChannelAccepted();
+            handler.stream.requestBytes(1);
+            var recvChunk = safePoll(handler.recvChunks);
+            assertTrue(recvChunk.isLast);
+            assertEquals(0, recvChunk.chunk.length());
+            recvChunk.chunk.close();
+        }
+    }
+
+    // ensures content integrity, no loses and re-order
+    public void testReceiveAllChunks() throws Exception {
+        try (var ctx = setupClientCtx()) {
+            var dataSize = randomIntBetween(1024, 10 * 1024 * 1024);
+            var sendData = Unpooled.wrappedBuffer(randomByteArrayOfLength(dataSize));
+            sendData.retain();
+            ctx.clientChannel.writeAndFlush(fullRequest(sendData));
+
+            var handler = ctx.awaitRestChannelAccepted();
+
+            var gotAllChunks = false;
+            var recvData = Unpooled.buffer(dataSize);
+            while (gotAllChunks == false) {
+                handler.stream.requestBytes(randomIntBetween(1024, 64 * 1024));
+                var recvChunk = safePoll(handler.recvChunks);
+                try (recvChunk.chunk) {
+                    if (recvChunk.isLast) {
+                        gotAllChunks = true;
+                    }
+                    recvData.writeBytes(BytesReference.toBytes(recvChunk.chunk));
+                }
+            }
+            assertEquals(sendData, recvData);
+        }
+    }
+
+    // ensure that client's socket is no longer writable if server not consuming data
+    public void testClientBackpressure() throws Exception {
```

Comment:
I think it would be useful to have test cases for exceptional scenarios such as client cancellation, server cancellation (or dropping data or closing channel) etc.

Reply:
Added 2 more tests `testClientConnectionCloseMidStream` and `testServerCloseConnectionMidStream`. Also made `testEmptyContent` and `testReceiveAllChunks` to send and receive random number of back-to-back requests rather than 1.